packages/core/src/input/InputManager.ts

```diff
@@ -0,0 +1,320 @@
+  }
+
+  /**
+   * Get the Entity to which the ray is cast.
+   * @param x - The X coordinate of the pointer on the screen, specified in normalized
+   * @param y - The Y coordinate of the pointer on the screen, specified in normalized
+   * @returns The Entity to which the ray is cast
+   */
+  private _pointerRaycast(x: number, y: number): Entity {
+    const { _engine: engine, _canvas: canvas } = this;
+    /** Convert screen coordinates to viewport coordinates. */
+    x = (x - canvas.offsetLeft) / canvas.clientWidth;
+    y = (y - canvas.offsetTop) / canvas.clientHeight;
+    const cameras = engine.sceneManager.activeScene._activeCameras;
+    for (let i = cameras.length - 1; i >= 0; i--) {
+      const camera = cameras[i];
+      if (!camera.enabled || camera.renderTarget) {
+        continue;
+      }
+      const { x: vpX, y: vpY, z: vpW, w: vpH } = camera.viewport;
+      if (x >= vpX && y >= vpY && x - vpX <= vpW && y - vpY <= vpH) {
+        const { _tempHitResult, _tempPoint } = InputManager;
+        _tempPoint.setValue((x - vpX) / vpW, (y - vpY) / vpH);
+        // TODO: Only check which colliders have listened to the input.
+        return engine.physicsManager.raycast(
+          camera.viewportPointToRay(_tempPoint, InputManager._tempRay),
+          _tempHitResult
+        )
+          ? _tempHitResult.collider.entity
+          : null;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Update pointers info.
+   * @param type - Pointer change type
+   * @param pointerId - Pointer ID
+   * @param x - The x coordinate of Pointer
+   * @param y - The y coordinate of Pointer
+   */
+  private _changePointer(type: PointerOperationType, pointerId: number, x?: number, y?: number): void {
+    const {
+      _pointerIdToIndexMap: _pointerIdToIndex,
+      _pointerList: pointerList,
+      _activePointerCount: lastCount,
+      _input: input
+    } = this;
+    let idx = _pointerIdToIndex[pointerId];
+    switch (type) {
+      case PointerOperationType.Remove:
+        if (idx === undefined) {
+          return;
+        } else {
+          const nowCount = (this._activePointerCount = lastCount - 1);
+          if (nowCount != 0) {
+            const removedTouch = pointerList[idx];
+            if (idx !== lastCount) {
+              pointerList[idx] = pointerList[lastCount];
+              pointerList[lastCount] = removedTouch;
+            }
+            input.x = (input.x * lastCount - removedTouch.position.x) / nowCount;
+            input.y = (input.y * lastCount - removedTouch.position.y) / nowCount;
+          }
+          delete _pointerIdToIndex[pointerId];
+        }
+        break;
+      case PointerOperationType.Update:
+        if (idx === undefined) {
+          if (lastCount > 0 && !this._multiTouchEnabled) {
+            return;
+          }
+          let touch: Pointer;
+          if (!pointerList[lastCount]) {
+            touch = pointerList[lastCount] = new Pointer(pointerId, x, y, lastCount);
```

Comment:
I like `pointerId` range always from (0, support maxPointerCount),  You can test other engine design.

If you do like this, you can use the index in pointerList as  pointerId and don't need to modify it forever.
Another advantage is developer cache Pointer use pointerId is better,because thb pointerId is fixed. Now will cause  confusion  when developer use  `pointerId` cache `Pointer` , because you may change `Pointer`'s `pointerId` next frame.

Reply:
Done.